src/coreclr/nativeaot/System.Private.StackTraceMetadata/src/Internal/StackTraceMetadata/StackTraceMetadata.cs

```diff
@@ -196,9 +198,8 @@ public unsafe PerModuleMethodNameResolver(IntPtr moduleAddress)
                     _metadataReader = new MetadataReader(new IntPtr(metadataBlob), (int)metadataBlobSize);
 
                     // RVA to token map consists of pairs of integers (method RVA - token)
-                    int rvaToTokenMapEntryCount = (int)(rvaToTokenMapBlobSize / (2 * sizeof(int)));
-                    _methodRvaToTokenMap = new Dictionary<int, int>(rvaToTokenMapEntryCount);
-                    PopulateRvaToTokenMap(handle, (int *)rvaToTokenMapBlob, rvaToTokenMapEntryCount);
+                    _methodRvaToTokenMap = new Dictionary<int, StackTraceData>();
```

Comment:
What is going to be memory hit from allocating this map in a small ASP.NET app?

This map was not small before this change, and it is going to be several times more with this change. I am wondering whether we need to do something to make the RVA to method reverse mapping more pay-for-play at runtime.

Reply:
For a small ASP.NET app we're talking about 50,000 entries in this dictionary.

Previously, this would be about (4 * 2 * 50) 400 kB of data. With this change it's about (4 * 5 * 50) 1000 kB of data.

It only gets allocated the first time someone asks for textual stack traces. In this sense it's pay for play because if you don't use it, you don't have to pay this.

Reply:
We could e.g. replace the dictionary with a giant array of StackTraceData (adding the RVA as another field to it) and binary search in it instead of the dictionary to make this exactly 1 MB with no dictionary overheads.

Reply:
Dictionary cost per entry is: KEY, VALUE, hashCode (int32), next (int32), bucket (int32). 

Since we are not preallocating the dictionary of the right size with this change, we will end up with 50% larger dictionary than we actually need on average, and we will pay for the large object heap garbage allocated in the process of the array resizing. The GC may be able to reclaim the LOH garbage eventually, but it is going to be slow process at best (LOH is expensive to defrag).

- Before this change: 4 * 5 * 50 = 1MB
- After this change: 4 * 8 * 50 * 1.5 = 2.4MB + close to the same amount of garbage in LOH

If I got the math right, formatting stacktrace is going to make the working set jump by ~1MB before this change and by ~5MB after this change. 5MB jump in a working set of a small ASP.NET app sounds like a lot. I agree that it is sort of pay-for-play, but I expect that every ASP.NET app that runs for a bit longer is going to throw an exception and pay this cost to log the exception. Things fail intermittently in the cloud all the time.

> We could e.g. replace the dictionary with a giant array of StackTraceData (adding the RVA as another field to it)

Yes, I think it would be better.

I am wondering whether we can avoid indexing the stacktrace data at runtime. We need to index the stacktrace at runtime because we depend on native linker to layout the managed code section. When we are compiling as single file (the only mode that we fully support), can we layout the managed code section in the compiler and create efficient lookup structure for things like the stack trace data in image and avoid the indexing of the stacktrace data at runtime?

Reply:
> When we are compiling as single file (the only mode that we fully support), can we layout the managed code section in the compiler and create efficient lookup structure for things like the stack trace data in image and avoid the indexing of the stacktrace data at runtime?

We could certainly do that and place the data in a hashtable hashed by the RVA, or sort by RVA. But it will come at a size cost too. The scheme in this PR relies on the methods being grouped in the mapping table by owning type/name/signature so that we don't need to duplicate this information. We'd have to start duplicating this again if we organize by something else. I don't see any way around that.

Maybe the best fix for now would be to just replace the dictionary with an array and keep building this at runtime?

We have a similar problem for `Delegate.GetMethodInfo` that also rebuilds a more efficient reverse lookup data structure from a static hashtable that is organized in a different direction. That one is mostly mitigated by the fact that we try not to have too many methods visible from reflection.

We could also try reducing the amount of methods in the stack trace mapping table. .NET Native had a blacklist of things that should never be present for example. Basically because most of the runtime is written in managed code, we can make stack traces for it, but CoreCLR/Mono would not have that data because it's native code.

Reply:
I guess we also have some spare bits in the `StackTraceDataCommand` where we could stash information about how big the previous record is. If we have that, we can keep the existing dictionary as `<int, int>` and the second `int` would be just an offset into the blob. When stack trace data is needed, we would look at the record on that offset and if it's not complete (e.g. owning type is missing), we would seek backwards to find the previous record that has it.

Reply:
> We could certainly do that and place the data in a hashtable hashed by the RVA

We can select a bucket using top bits of relative code offset in the managed code section (e.g. top 19 bits - bucket size 8kB) and then do a linear search within the bucket. The symbol information for all method in the bucket can be encoded using the delta encoding like what you have proposed here. If we do the layout in the compiler, the delta encoding can be applied even on method addresses so we would typically only need 1 or 2 bytes to encode next method address instead of full 4 bytes that we need today. The methods from the same type are often placed next to each other, so we would still get some gains from the delta encoding of the names. My guess is that the binary footprint would not be than much worse compared to what you are proposing here, except that the runtime behavior would scale much better.

> the best fix for now would be to just replace the dictionary with an array and keep building this at runtime?

Yes, it would be nice to avoid regressing the current runtime behavior (that is not great).